File: pa6-cky-v3/python/io/NumberRangeFileFilter.py
import os

import ling.Trees as Trees

class NumberRangeFileFilter:
    """
    Class to use as filter for files (by file number).
    """
    def __init__(self, extension, low_filenum, high_filenum, recurse):
        self.i = -1
        self.high_filenum = high_filenum
        self.low_filenum = low_filenum
        self.extension = extension
        self.recurse = recurse


    def accept(self, pathname):
        if os.path.isdir(pathname):
            return self.recurse
        name = os.path.basename(pathname)
        if not name.endswith(self.extension):
            return False
        last_num_index = self.get_last_number_index(name)
        if last_num_index == -1:
            return False
        num_end_loc = last_num_index + 1
        num_start_loc = self.get_last_non_number_index(name, last_num_index) + 1
        file_num = int(name[num_start_loc:num_end_loc])
    

File: pa6-cky-v3/java/src/nlpclass/io/NumberRangeFileFilter.java

package nlpclass.io;

import java.io.FileFilter;
import java.io.File;

/**
 * Accepts files based on the (last) number in their filename.  Optionally restricts based on extensions, as well.
 *
 * @author Dan Klein
 */
class NumberRangeFileFilter implements FileFilter {
  int highFileNum;
  int lowFileNum;
  String extension;
  boolean recurse;
  int i;

  public boolean accept(File pathname) {
    if (pathname.isDirectory())
      return recurse;
    String name = pathname.getName();
    if (! name.endsWith(extension))
      return false;
    int lastNumberIndex = getLastNumberIndex(name);
    if (lastNumberIndex == -1)
      return false;
    int numEndLoc = lastNumberIndex+1;
    int numStartLoc = getLastNonNumberIndex(name, lastNumberIndex)+1;
    int fileNum = Integer.parseInt(name.substring(numStartLoc, numEndLoc));
    if (fileNum >= lowFileNum && fileNum <= highFileNum)
      return true;
    return false;
  }
  
  public boolean acceptSequential(File pathname) {
	if (pathname.isDirectory())
	  return recurse;
	String name = pathname.getName();
    if (! name.endsWith(extension))
      return false;
    ++i;
    return i >= lowFileNum && i <= highFileNum;
  }

  private int getLastNonNumberIndex(String name, int lastNumberIndex) {
    int index = lastNumberIndex - 1;
    while (index >= 0 && Character.isDigit(name.charAt(index))) {
      index--;
    }
    if (index < -1)
      return -1;
    return index;
  }

  private int getLastNumberIndex(String name) {
    int index = name.length()-1;
    while (index >= 0 && ! Character.isDigit(name.charAt(index))) {
      index--;
    }
    return index;
  }

  public NumberRangeFileFilter() {
  }
  
  public NumberRangeFileFilter(String extension, int lowFileNum, int highFileNum, boolean recurse) {
	this.i = -1; // initialize file counter to zero
    this.highFileNum = highFileNum;
    this.lowFileNum = lowFileNum;
    this.extension = extension;
    this.recurse = recurse;
  }
}